src/cargo/core/compiler/fingerprint.rs

```diff
@@ -1779,28 +1843,155 @@ pub fn translate_dep_info(
+                    | ((ret[1] as usize) << 8)
+                    | ((ret[2] as usize) << 16)
+                    | ((ret[3] as usize) << 24),
+            )
+        }
+
+        fn read_u8(bytes: &mut &[u8]) -> Option<u8> {
+            let ret = *bytes.get(0)?;
+            *bytes = &bytes[1..];
+            Some(ret)
+        }
+
+        fn read_bytes<'a>(bytes: &mut &'a [u8]) -> Option<&'a [u8]> {
+            let n = read_usize(bytes)? as usize;
+            let ret = bytes.get(..n)?;
+            *bytes = &bytes[n..];
+            Some(ret)
+        }
+    }
+
+    fn serialize(&self) -> CargoResult<Vec<u8>> {
+        let mut ret = Vec::new();
+        let dst = &mut ret;
+        write_usize(dst, self.files.len());
+        for (ty, file) in self.files.iter() {
+            match ty {
+                DepInfoPathType::PackageRootRelative => dst.push(0),
+                DepInfoPathType::TargetRootRelative => dst.push(1),
+            }
+            write_bytes(dst, util::path2bytes(file)?);
+        }
+
+        write_usize(dst, self.env.len());
+        for (key, val) in self.env.iter() {
+            write_bytes(dst, key);
+            match val {
+                None => dst.push(0),
+                Some(val) => {
+                    dst.push(1);
+                    write_bytes(dst, val);
+                }
+            }
+        }
+        return Ok(ret);
+
+        fn write_bytes(dst: &mut Vec<u8>, val: impl AsRef<[u8]>) {
+            let val = val.as_ref();
+            write_usize(dst, val.len());
+            dst.extend_from_slice(val);
+        }
+
+        fn write_usize(dst: &mut Vec<u8>, val: usize) {
+            dst.push(val as u8);
+            dst.push((val >> 8) as u8);
+            dst.push((val >> 16) as u8);
+            dst.push((val >> 24) as u8);
+            assert!(val >> 32 == 0);
+        }
+    }
+}
+
 /// Parse the `.d` dep-info file generated by rustc.
-///
-/// Result is a Vec of `(target, prerequisites)` tuples where `target` is the
-/// rule name, and `prerequisites` is a list of files that it depends on.
-pub fn parse_rustc_dep_info(rustc_dep_info: &Path) -> CargoResult<Vec<(String, Vec<String>)>> {
+pub fn parse_rustc_dep_info(rustc_dep_info: &Path) -> CargoResult<RustcDepInfo> {
     let contents = paths::read(rustc_dep_info)?;
-    contents
-        .lines()
-        .filter_map(|l| l.find(": ").map(|i| (l, i)))
-        .map(|(line, pos)| {
-            let target = &line[..pos];
+    let mut ret = RustcDepInfo::default();
+    let mut found_deps = false;
+
+    for line in contents.lines() {
+        let env_dep_prefix = "# env-dep:";
+        if line.starts_with(env_dep_prefix) {
+            let rest = &line[env_dep_prefix.len()..];
+            let mut parts = rest.splitn(2, '=');
+            let env_var = match parts.next() {
+                Some(s) => s.to_string(),
+                None => continue,
+            };
+            let env_val = parts.next().map(|s| s.to_string());
```

Comment:
I think this needs to support escaping. Otherwise, the fingerprint will always be considered dirty. I'd also like to see a test for that.  Alternatively, cargo could just re-fetch the value directly from the env instead of reading from the file.